### ISSUE TITLE
table have more than one primary_key can not use  on_conflict(diesel::dsl::DuplicatedKeys) 

### ISSUE THREAD

**Issue by Caisin** [diesel-rs/diesel#3872]
## Setup
schema.rs
```
diesel::table! {
    sys_api (path, method) {
        #[max_length = 255]
        path -> Varchar,
        #[max_length = 50]
        method -> Varchar,
        #[max_length = 50]
        name -> Varchar,
        #[max_length = 50]
        scope -> Varchar,
        #[max_length = 10]
        state -> Varchar,
    }
}
```
models.rs
```
#[derive(
    Clone,
    Debug,
    Serialize,
    Deserialize,
    Default,
    PartialEq,
    Queryable,
    Insertable,
    Identifiable,
    AsChangeset,
)]
#[diesel(table_name = sys_api)]
#[diesel(primary_key(path,method))]
pub struct SysApi {
    pub id: u64,
    pub path: String,
    pub method: String,
    pub name: String,
    pub scope: String,
    pub state: String,
}
```
### Versions

- **Rust:**1.76
- **Diesel:**2.14
- **Database:**mysql
- **Operating System**windows

### Feature Flags

- **diesel:**

## Problem Description
table have more than one primary_key can not use  on_conflict(diesel::dsl::DuplicatedKeys) 

### What is the actual output?
```
 Diesel,
     |     ^^^^^^ the trait `QueryFragment<diesel::mysql::Mysql, mysql::backend::MysqlOnConflictClause>` is not implemented for `query_builder::upsert::on_conflict_clause::OnConflictValues<query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::path, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::method, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::name, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::scope, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::state, expression::bound::Bound<Text, &std::string::String>>>), schema::sys_api::table>, query_builder::upsert::on_conflict_target::ConflictTarget<DuplicatedKeys>, query_builder::upsert::on_conflict_actions::DoUpdate<(query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::name>, SqlLiteral<Text>>, query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::scope>, SqlLiteral<Text>>, query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::state>, SqlLiteral<Text>>), schema::sys_api::table>>`
     |
     = help: the following other types implement trait `QueryFragment<DB, SP>`:
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<diesel::mysql::Mysql, mysql::backend::MysqlOnConflictClause>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<DB>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<DB, SD>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action, query_builder::where_clause::WhereClause<Expr>> as QueryFragment<DB>>
     = note: required for `OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, Bound<Text, &String>>>, ..., ..., ..., ...), ...>, ..., ...>` to implement `QueryFragment<diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-17862108743910028546.txt'
     = note: 1 redundant requirement hidden
     = note: required for `InsertStatement<table, OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, ...>>, ..., ..., ..., ...), ...>, ..., ...>>` to implement `QueryFragment<diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-2431980552689726686.txt'
     = note: required for `InsertStatement<table, OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, ...>>, ..., ..., ..., ...), ...>, ..., ...>>` to implement `ExecuteDsl<PooledConnection<ConnectionManager<MysqlConnection>>, diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-2431980552689726686.txt'
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> D:\code\rust\diesel\diesel\src\query_dsl\mod.rs:1431:15
     |
1428 |     fn execute(self, conn: &mut Conn) -> QueryResult<usize>
     |        ------- required by a bound in this associated function
...
1431 |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::execute`
     = note: this error originates in the derive macro `Diesel` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
Thanks for opening this issue. Please provide all the relevant code that is necessary to reproduce your error message, as otherwise it's hard/impossible for anyone to see what's wrong here.  Also please provide the complete error message as emitted by rustc. Based on the provided part of the error message it seems to miss relevant information (like line information and code snippet of the actual error.)

**Comment by Caisin** [comment]
# FAIL code
**schema.rs**
```
diesel::table! {
    sys_api (path, method) {
        #[max_length = 255]
        path -> Varchar,
        #[max_length = 50]
        method -> Varchar,
        #[max_length = 50]
        name -> Varchar,
        #[max_length = 50]
        scope -> Varchar,
        #[max_length = 10]
        state -> Varchar,
    }
}
```
**models.rs**
```
#[derive(
    Clone,
    Debug,
    Serialize,
    Deserialize,
    Default,
    PartialEq,
    Queryable,
    Insertable,
    Identifiable,
    AsChangeset,
)]
#[diesel(table_name = sys_api)]
#[diesel(primary_key(path,method))]
pub struct SysApi {
    pub path: String,
    pub method: String,
    pub name: String,
    pub scope: String,
    pub state: String,
}
```
**insert into code**
```
let v = SysApi::default();
diesel::insert_into(sys_api)
            .values(&v)
            .on_conflict(diesel::dsl::DuplicatedKeys)
            .do_update()
            .set(&v)
            .execute(con)?;
```

**error log** 
```

Diesel,
     |     ^^^^^^ the trait `QueryFragment<diesel::mysql::Mysql, mysql::backend::MysqlOnConflictClause>` is not implemented for `query_builder::upsert::on_conflict_clause::OnConflictValues<query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::path, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::method, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::name, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::scope, expression::bound::Bound<Text, &std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<schema::sys_api::columns::state, expression::bound::Bound<Text, &std::string::String>>>), schema::sys_api::table>, query_builder::upsert::on_conflict_target::ConflictTarget<DuplicatedKeys>, query_builder::upsert::on_conflict_actions::DoUpdate<(query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::name>, SqlLiteral<Text>>, query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::scope>, SqlLiteral<Text>>, query_builder::update_statement::changeset::Assign<query_builder::update_statement::changeset::ColumnWrapperForUpdate<schema::sys_api::columns::state>, SqlLiteral<Text>>), schema::sys_api::table>>`
     |
     = help: the following other types implement trait `QueryFragment<DB, SP>`:
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<diesel::mysql::Mysql, mysql::backend::MysqlOnConflictClause>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<DB>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action> as QueryFragment<DB, SD>>
               <query_builder::upsert::on_conflict_clause::OnConflictValues<Values, Target, Action, query_builder::where_clause::WhereClause<Expr>> as QueryFragment<DB>>
     = note: required for `OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, Bound<Text, &String>>>, ..., ..., ..., ...), ...>, ..., ...>` to implement `QueryFragment<diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-17862108743910028546.txt'
     = note: 1 redundant requirement hidden
     = note: required for `InsertStatement<table, OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, ...>>, ..., ..., ..., ...), ...>, ..., ...>>` to implement `QueryFragment<diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-2431980552689726686.txt'
     = note: required for `InsertStatement<table, OnConflictValues<ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<path, ...>>, ..., ..., ..., ...), ...>, ..., ...>>` to implement `ExecuteDsl<PooledConnection<ConnectionManager<MysqlConnection>>, diesel::mysql::Mysql>`
     = note: the full type name has been written to 'D:\code\rust\cx\target\debug\deps\cx_biz_sys-6a229df924011786.long-type-2431980552689726686.txt'
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> D:\code\rust\diesel\diesel\src\query_dsl\mod.rs:1431:15
     |
1428 |     fn execute(self, conn: &mut Conn) -> QueryResult<usize>
     |        ------- required by a bound in this associated function
...
1431 |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::execute`
     = note: this error originates in the derive macro `Diesel` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```
## when id use one Field primary_key this code  cargo check pass

# Pass Code

**schema.rs**
```
diesel::table! {
    sys_api (id) {
        id -> Unsigned<BigInt>,
        #[max_length = 255]
        path -> Varchar,
        #[max_length = 50]
        method -> Varchar,
        #[max_length = 50]
        name -> Varchar,
        #[max_length = 50]
        scope -> Varchar,
        #[max_length = 10]
        state -> Varchar,
    }
}
```
**models.rs**
```
#[derive(
    Clone,
    Debug,
    Serialize,
    Deserialize,
    Default,
    PartialEq,
    Queryable,
    Insertable,
    Identifiable,
    AsChangeset,
)]
#[diesel(table_name = sys_api)]
#[diesel(primary_key(id))]
pub struct SysApi {
    pub id: u64,
    pub path: String,
    pub method: String,
    pub name: String,
    pub scope: String,
    pub state: String,
}
```
**insert into code**
```
let v = SysApi::default();
diesel::insert_into(sys_api)
            .values(&v)
            .on_conflict(diesel::dsl::DuplicatedKeys)
            .do_update()
            .set(&v)
            .execute(con)?;
```

**Comment by Caisin** [comment]
@weiznich please help confirm again

**Comment by weiznich** [comment]
As this is a project run in my free time I can only work on stuff if I have time. That usually means that it can take several days/weeks/month until you hear back from me about something. Pinging me usually puts such things down on my priority list.

**Comment by weiznich** [comment]
I've added #3877 to fix this issue. Please check if that resolves your problem.

**Comment by weiznich** [comment]
Closed by #3877

**Comment by Caisin** [comment]
@weiznich 3q verify much
